### ISSUE TITLE
Fancy reprs

### ISSUE THREAD

**Issue by benbovy** [pydata/xarray#8171]
### What is your issue?

In Xarray we already have the plain-text and html reprs, which is great.

Recently, I've tried [anywidget](https://anywidget.dev/) and I think that it has potential to overcome some of the limitations of the current repr and possibly go well beyond it.

The main advantages of anywidget:

- it is broadly compatible with jupyter-like front-ends (Jupyterlab, notebook, vscode, colab, etc.), although I haven't tested it myself on all those front-ends yet.
- it is super easy to get started: almost no project setup (build, packaging) is required before experimenting with it, although it still requires writing Javascript / HTML / CSS, etc..

I don't think we should replace the current html repr (it is still useful to have a basic, pure HTML/CSS version), but having a new widget could improve some aspects like not including the whole CSS each time an object repr is displayed, removing some HTML/CSS hacks... and actually has much more potential since we would have the whole javascript ecosystem at our fingertips (quick plots, etc.). Also bi-directional communication with Python is possible.

I'm opening this issue to brainstorm about what would be nice to have in widget-based Xarray reprs:

- fancy hover effects (e.g., highlight all variables sharing common dimensions, coordinates sharing a common index, etc.)
- more icons next to each variable reprs (attributes, array repr, quick plot? quick map?)
- ... ?

cc @pydata/xarray

**Comment by shoyer** [comment]
This sounds exciting! Assuming the minimal dependencies are not too painful to add, this sounds like a nice way to add optional dynamic effects.

Do you understand how they solve JavaScript sandboxing issue? Or maybe that only works on particular notebook implementations?

**Comment by TomNicholas** [comment]
Great idea. One thing this could be useful for is interactively visualizing the structure of `DataTree` objects. https://github.com/xarray-contrib/datatree/issues/207

**Comment by benbovy** [comment]
Perhaps this would better be implemented in a third party package that xarray would optionally depend on, and fallback to the pure html/css repr if it is not installed.

Anywidget relies on standard ES modules, which I think makes it more easily front-end agnostic.

**Comment by benbovy** [comment]
@TomNicholas yes we have just talked about that with @keewis !

**Comment by benbovy** [comment]
I started experimenting with this in https://github.com/benbovy/xarray-fancy-repr

**Comment by benbovy** [comment]
> started experimenting with this in https://github.com/benbovy/xarray-fancy-repr

First release is now available on PyPI: `pip install xarray-fancy-repr` !

It has reprs for Dataset, DataArray, Coordinates and Variable (copied from the Xarray html reprs). For now it also implements a basic search tool to filter variables and attributes (by name, dimension and/or attributes). I opened a few issues for other ideas.

![xarray-fancy-repr2](https://github.com/pydata/xarray/assets/4160723/c376107e-f62f-4680-8dba-b44517260ddf)

**Comment by dcherian** [comment]
Now that's pretty cool!

**Comment by rabernat** [comment]
One really cool direction to take this would be to add automatic visualization of the data variables. Like, rather than just showing the dask repr or a bunch of numbers, could we make an interactive plot appear?

**Comment by benbovy** [comment]
> One really cool direction to take this would be to add automatic visualization of the data variables. Like, rather than just showing the dask repr or a bunch of numbers, could we make an interactive plot appear?

Yes there is room for adding another few icons next to each variable inline repr for that.

**Comment by max-sixty** [comment]
Very cool!
